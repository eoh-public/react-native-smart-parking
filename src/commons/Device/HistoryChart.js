import React, { memo, useCallback, useState } from 'react';
import { StyleSheet, View, Dimensions } from 'react-native';
import moment from 'moment';
import t from 'i18n';

import { Colors } from '../../configs';
import Text from '../../commons/Text';
import DateTimeRangeChange from '../DateTimeRangeChange';
import HorizontalBarChart from './HorizontalBarChart';
import DateTimePickerModal from 'react-native-modal-datetime-picker';
import LinearChart from './LinearChart';

export const dateTimeType = {
  date: 'date',
  time: 'time',
  dateTime: 'datetime',
};

const { width } = Dimensions.get('window');

const HistoryChart = memo(
  ({
    datas,
    unit,
    style,
    formatType,
    startDate,
    setEndDate,
    setStartDate,
    configuration,
  }) => {
    const dateNow = moment().valueOf();
    const [chartOptions, setChartOptions] = useState({
      index: -1,
      showAll: true,
    });
    const [eventPicker, setEventPicker] = useState({
      currentChangeTime: '',
      showModalEnd: false,
      showModalStart: false,
      startTime: startDate ? startDate : moment().subtract(1, 'day').valueOf(),
      endTime: dateNow,
    });
    const onStart = useCallback(() => {
      setEventPicker({
        ...eventPicker,
        currentChangeTime: 'start',
        showModalStart: true,
        showModalEnd: false,
      });
    }, [eventPicker]);
    const onEnd = useCallback(() => {
      setEventPicker({
        ...eventPicker,
        currentChangeTime: 'end',
        show: true,
        showModalStart: false,
        showModalEnd: true,
      });
    }, [eventPicker]);
    const onConfirmStart = (date) => {
      if (typeof date === 'number') {
        onCancel();
        return;
      }
      if (moment(date).valueOf() < eventPicker.endTime) {
        setEventPicker({
          ...eventPicker,
          currentChangeTime: 'start',
          showModalStart: false,
          startTime: moment(date).valueOf(),
        });
      } else {
        setEventPicker({
          ...eventPicker,
          currentChangeTime: 'start',
          showModalStart: false,
          startTime: moment(date).valueOf(),
          endTime: moment(date).add(1, 'day').valueOf(),
        });
      }

      setStartDate(moment(date).valueOf());
    };

    const onConfirmEnd = (date) => {
      if (typeof date === 'number') {
        onCancel();
        return;
      }
      if (moment(date).valueOf() >= eventPicker.startTime) {
        setEventPicker({
          ...eventPicker,
          currentChangeTime: 'start',
          showModalEnd: false,
          endTime: moment(date).valueOf(),
        });
      } else {
        setEventPicker({
          ...eventPicker,
          currentChangeTime: 'start',
          showModalEnd: false,
          endTime: moment(date).valueOf(),
          startTime: moment(date).subtract(1, 'day').valueOf(),
        });
      }
      setEndDate(moment(date).valueOf());
    };

    const onCancel = useCallback(() => {
      setEventPicker({
        ...eventPicker,
        currentChangeTime: '',
        showModalEnd: false,
        showModalStart: false,
      });
    }, [eventPicker]);
    const onShowOneChart = useCallback(
      (index) => {
        if (index === chartOptions.index) {
          setChartOptions({ index, showAll: !chartOptions.showAll });
        } else {
          setChartOptions({ index, showAll: false });
        }
      },
      [chartOptions]
    );
    const renderChart = useCallback(() => {
      if (configuration.type === 'line_chart') {
        return (
          <View style={styles.chartContainer}>
            <View style={[styles.chart, { width: width }]}>
              <LinearChart datas={datas} />
            </View>
          </View>
        );
      }
      if (configuration.type === 'horizontal_bar_chart') {
        return <HorizontalBarChart unit={unit} datas={datas} />;
      }
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [configuration, datas, chartOptions, onShowOneChart]);

    return (
      <View style={style}>
        <View style={styles.historyView}>
          <Text size={20} semibold color={Colors.Gray9}>
            {t('history')}
          </Text>
          <DateTimeRangeChange
            startTime={eventPicker.startTime}
            onStart={onStart}
            onEnd={onEnd}
            endTime={eventPicker.endTime}
            date={dateNow}
            formatType={formatType}
          />
        </View>
        {renderChart()}
        <DateTimePickerModal
          isVisible={eventPicker.showModalStart}
          date={eventPicker.startTime}
          mode={formatType || 'datetime'}
          onConfirm={onConfirmStart}
          onCancel={onCancel}
          display="spinner"
        />
        <DateTimePickerModal
          isVisible={eventPicker.showModalEnd}
          date={eventPicker.endTime}
          mode={formatType || 'datetime'}
          onConfirm={onConfirmEnd}
          onCancel={onCancel}
          display="spinner"
        />
      </View>
    );
  }
);

export default HistoryChart;

const styles = StyleSheet.create({
  historyView: {
    paddingTop: 16,
  },
  chartInfo: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  chartContainer: {
    marginLeft: -8,
  },
  chart: {
    height: 300,
  },
});
