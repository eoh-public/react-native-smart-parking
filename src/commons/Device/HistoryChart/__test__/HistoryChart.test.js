import React from 'react';
import renderer from 'react-test-renderer';
import HistoryChart from '../../HistoryChart';

describe('Test HistoryChart', () => {
  beforeAll(() => {
    Date.now = jest.fn(() => new Date('2021-01-24T12:00:00.000Z'));
  });
  test('render HistoryChart line_chart', () => {
    const datas = [
      {
        id: 1,
        title: 'title',
        color: 'red',
        data: [
          [1, 2],
          [2, 3],
        ],
      },
    ];
    const render = renderer
      .create(
        <HistoryChart
          configuration={{ type: 'line_chart' }}
          datas={datas}
          setStartDate={Date.now()}
          setEndDate={Date.now()}
        />
      )
      .toJSON();
    expect(render).toMatchSnapshot();
  });
  test('render HistoryChart horizontal_bar_chart', () => {
    const datas = [
      {
        id: 1,
        title: 'title',
        color: 'red',
        data: [
          [1, 2],
          [2, 3],
        ],
      },
    ];
    const render = renderer
      .create(
        <HistoryChart
          configuration={{ type: 'horizontal_bar_chart' }}
          datas={datas}
          setStartDate={Date.now}
          setEndDate={Date.now}
        />
      )
      .toJSON();
    expect(render).toMatchSnapshot();
  });
});
