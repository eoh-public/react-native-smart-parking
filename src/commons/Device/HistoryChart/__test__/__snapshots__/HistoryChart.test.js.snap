// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test HistoryChart render HistoryChart horizontal_bar_chart 1`] = `
<View>
  <View
    style={
      Object {
        "paddingTop": 16,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#262626",
            "fontFamily": "SFProDisplay-Semibold",
            "fontSize": 20,
            "lineHeight": 20,
            "textAlign": "left",
            "textDecorationLine": "none",
          },
          Object {},
        ]
      }
    >
      Lịch sử
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginTop": 16,
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#8C8C8C",
              "fontFamily": "SFProDisplay-Regular",
              "fontSize": 12,
              "lineHeight": 12,
              "textAlign": "left",
              "textDecorationLine": "none",
            },
            Object {},
          ]
        }
      >
        từ
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderColor": "#D9D9D9",
            "borderRadius": 2,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginHorizontal": 8,
            "opacity": 1,
            "paddingLeft": 12,
            "paddingRight": 8,
            "paddingVertical": 8,
            "width": 124,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#595959",
                "fontFamily": "SFProDisplay-Regular",
                "fontSize": 12,
                "lineHeight": 20,
                "textAlign": "left",
                "textDecorationLine": "none",
              },
              Object {
                "marginRight": 4,
              },
            ]
          }
        >
          23.01.21 12:00
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#595959",
                "fontFamily": "antfill",
                "fontSize": 16,
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              undefined,
            ]
          }
        >
          
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#8C8C8C",
              "fontFamily": "SFProDisplay-Regular",
              "fontSize": 12,
              "lineHeight": 12,
              "textAlign": "left",
              "textDecorationLine": "none",
            },
            Object {},
          ]
        }
      >
        đến
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderColor": "#D9D9D9",
            "borderRadius": 2,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginHorizontal": 8,
            "opacity": 1,
            "paddingLeft": 12,
            "paddingRight": 8,
            "paddingVertical": 8,
            "width": 124,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#595959",
                "fontFamily": "SFProDisplay-Regular",
                "fontSize": 12,
                "lineHeight": 20,
                "textAlign": "left",
                "textDecorationLine": "none",
              },
              Object {
                "marginRight": 4,
              },
            ]
          }
        >
          24.01.21 12:00
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#595959",
                "fontFamily": "antfill",
                "fontSize": 16,
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              undefined,
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
          "justifyContent": "center",
          "marginRight": 24,
          "marginTop": 5,
        },
        Object {
          "height": 110,
        },
      ]
    }
  >
    <View />
  </View>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    isVisible={false}
    onHide={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
  </Modal>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    isVisible={false}
    onHide={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
  </Modal>
</View>
`;

exports[`Test HistoryChart render HistoryChart line_chart 1`] = `
<View>
  <View
    style={
      Object {
        "paddingTop": 16,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#262626",
            "fontFamily": "SFProDisplay-Semibold",
            "fontSize": 20,
            "lineHeight": 20,
            "textAlign": "left",
            "textDecorationLine": "none",
          },
          Object {},
        ]
      }
    >
      Lịch sử
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginTop": 16,
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#8C8C8C",
              "fontFamily": "SFProDisplay-Regular",
              "fontSize": 12,
              "lineHeight": 12,
              "textAlign": "left",
              "textDecorationLine": "none",
            },
            Object {},
          ]
        }
      >
        từ
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderColor": "#D9D9D9",
            "borderRadius": 2,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginHorizontal": 8,
            "opacity": 1,
            "paddingLeft": 12,
            "paddingRight": 8,
            "paddingVertical": 8,
            "width": 124,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#595959",
                "fontFamily": "SFProDisplay-Regular",
                "fontSize": 12,
                "lineHeight": 20,
                "textAlign": "left",
                "textDecorationLine": "none",
              },
              Object {
                "marginRight": 4,
              },
            ]
          }
        >
          23.01.21 12:00
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#595959",
                "fontFamily": "antfill",
                "fontSize": 16,
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              undefined,
            ]
          }
        >
          
        </Text>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#8C8C8C",
              "fontFamily": "SFProDisplay-Regular",
              "fontSize": 12,
              "lineHeight": 12,
              "textAlign": "left",
              "textDecorationLine": "none",
            },
            Object {},
          ]
        }
      >
        đến
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderColor": "#D9D9D9",
            "borderRadius": 2,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginHorizontal": 8,
            "opacity": 1,
            "paddingLeft": 12,
            "paddingRight": 8,
            "paddingVertical": 8,
            "width": 124,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#595959",
                "fontFamily": "SFProDisplay-Regular",
                "fontSize": 12,
                "lineHeight": 20,
                "textAlign": "left",
                "textDecorationLine": "none",
              },
              Object {
                "marginRight": 4,
              },
            ]
          }
        >
          24.01.21 12:00
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#595959",
                "fontFamily": "antfill",
                "fontSize": 16,
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              undefined,
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "marginLeft": -8,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "height": 300,
          },
          Object {
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "flex": 1,
            "justifyContent": "center",
            "marginRight": 24,
            "marginTop": 24,
          }
        }
      >
        <View />
      </View>
    </View>
  </View>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    isVisible={false}
    onHide={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
  </Modal>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    isVisible={false}
    onHide={[Function]}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "height": 1334,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "width": 750,
        }
      }
    />
  </Modal>
</View>
`;
